### PR TITLE
Migrate Check Modern language

### PR DESCRIPTION
**Pull request by D074020**

### CODE REVIEW

File: src/checks/#cc4a#modern_language.clas.abap [hunk 1]
```diff
@@ -0,0 +1,1157 @@
...
+                VALUE(from_idx) TYPE i OPTIONAL
+                VALUE(to_idx)   TYPE i OPTIONAL
+      CHANGING  result          TYPE string.
+    METHODS get_receiving_infos
+      IMPORTING
+        tokens        TYPE if_ci_atc_source_code_provider=>ty_tokens
+      EXPORTING
+        receiving_idx TYPE i
+        end_idx       TYPE i
+        result_line   TYPE string.
```
Comment (bjoern-jueliger-sap) [inline src/checks/#cc4a#modern_language.clas.abap:121]: We should adhere to our own [guidelines](https://github.com/SAP/styleguides/blob/main/clean-abap/CleanABAP.md#prefer-returning-to-exporting): Use RETURNING where possible, and here's it's easily possible - just return a structure.

File: src/checks/#cc4a#modern_language.clas.abap [hunk 2]
```diff
@@ -0,0 +1,1157 @@
...
+    CONSTANTS c_qf_exporting_receiving TYPE cl_ci_atc_quickfixes=>ty_quickfix_code VALUE 'QF_EXP_REC'.
+    CONSTANTS c_qf_text_assembly TYPE cl_ci_atc_quickfixes=>ty_quickfix_code VALUE 'QF_TEXTASM'.
+    CONSTANTS c_ps_deprecated_key TYPE string VALUE 'DEPRECATED_KEY'.
+    CONSTANTS c_ps_line_exists TYPE string VALUE 'PREF_LINE_EX'.
+    CONSTANTS c_ps_prefer_new TYPE string VALUE 'PREF_NEW'.
+    CONSTANTS c_ps_call_method TYPE string VALUE 'CALL_METH_USAGE'.
+    CONSTANTS c_ps_method_exporting TYPE string VALUE 'OPTL_EXP'.
+    CONSTANTS c_ps_exporting_receiving TYPE string VALUE 'RECEIVING_USAGE'.
+    CONSTANTS c_ps_text_assembly TYPE string VALUE 'TEXT_ASSEMBLY'.
+    CONSTANTS c_max_line_length TYPE i VALUE 255.
```
Comment (bjoern-jueliger-sap) [inline src/checks/#cc4a#modern_language.clas.abap:34]: 1. Please don't use hungarian notation (cf. [style guide](https://github.com/SAP/styleguides/blob/main/clean-abap/CleanABAP.md#avoid-encodings-esp-hungarian-notation-and-prefixes))
2. Try to be consistent with the other checks structure their constants for finding codes and use structured constants for the message codes and quickfix codes like e.g. [here](https://github.com/SAP/code-pal-for-abap-cloud/blob/a28cadc0ef08896aa69bc5e9f04d84eb8527e5a1/src/checks/%23cc4a%23method_signature.clas.abap#L10)
3. Do all these constants need to be public? E.g. `c_max_line_length` doesn't seem useful to external callers (and if we need a constant for the ABAP line length its correct place is probably in `/CC4A/IF_ABAP_ANALYZER`).

File: src/checks/#cc4a#modern_language.clas.abap [hunk 3]
```diff
@@ -0,0 +1,1157 @@
...
+    DATA assistant_factory TYPE REF TO cl_ci_atc_assistant_factory.
+
+    METHODS analyze_procedure
+      IMPORTING procedure       TYPE if_ci_atc_source_code_provider=>ty_procedure
+      RETURNING VALUE(findings) TYPE if_ci_atc_check=>ty_findings.
+    METHODS analyze_move
+      IMPORTING procedure       TYPE if_ci_atc_source_code_provider=>ty_procedure
+                statement_index TYPE i
+      EXPORTING changed_line    TYPE string
+      RETURNING VALUE(findings) TYPE if_ci_atc_check=>ty_findings.
```
Comment (bjoern-jueliger-sap) [inline src/checks/#cc4a#modern_language.clas.abap:48]: [Don't mix exporting and returning parameters](https://github.com/SAP/styleguides/blob/main/clean-abap/CleanABAP.md#use-either-returning-or-exporting-or-changing-but-not-a-combination), return a structure instead.

File: src/checks/#cc4a#modern_language.clas.abap [hunk 4]
```diff
@@ -0,0 +1,1157 @@
...
+         statement_index = statement_index
+         code = c_code_line_exists
+         pseudo_comment = c_ps_line_exists
+      CHANGING findings = findings ).
+      RETURN.
+    ENDIF.
+    DATA(quickfixable) = abap_true.
+*    if statement-tokens[ key_idx + 2 ]-lexeme = '='.
+*      quickfixable = abap_false.
+*    endif.
```
Comment (bjoern-jueliger-sap) [inline src/checks/#cc4a#modern_language.clas.abap:391]: Please don't leave commented-out code around.

File: src/checks/#cc4a#modern_language.clas.abap [hunk 5]
```diff
@@ -0,0 +1,1157 @@
...
+      DATA(and_idx) = analyzer->find_clause_index( tokens = statement-tokens start_index = token_idx + 1 clause = 'AND' ).
+      IF and_idx = 0.
+        append_tokens( EXPORTING tokens = statement-tokens from_idx = token_idx + 1 CHANGING result = key ).
+        EXIT.
+      ELSE.
+        append_tokens( EXPORTING tokens = statement-tokens from_idx = token_idx + 1 to_idx = and_idx - 1 CHANGING result = key ).
+        token_idx = and_idx + 1.
+      ENDIF.
+    ENDDO.
+    DATA(quickfixable) = abap_true.
```
Comment (bjoern-jueliger-sap) [inline src/checks/#cc4a#modern_language.clas.abap:573]: This field is never read (according to SLIN)

File: src/checks/#cc4a#modern_language.clas.abap [hunk 6]
```diff
@@ -0,0 +1,1157 @@
...
+      add_finding(
+      EXPORTING
+         procedure = procedure
+         statement_index = statement_index
+         code = c_code_line_exists
+         pseudo_comment = c_ps_line_exists
+      CHANGING findings = findings ).
+    ENDIF.
+
+  ENDMETHOD.
```
Comment (bjoern-jueliger-sap) [inline src/checks/#cc4a#modern_language.clas.abap:517]: This method had 180 lines, any chance of splitting this up into more readable chunks?

File: src/checks/#cc4a#modern_language.clas.abap [hunk 7]
```diff
@@ -0,0 +1,1157 @@
...
+        append_tokens( EXPORTING tokens = statement-tokens from_idx = token_idx + 1 to_idx = and_idx - 1 CHANGING result = key ).
+        token_idx = and_idx + 1.
+      ENDIF.
+    ENDDO.
+    DATA(quickfixable) = abap_true.
+    LOOP AT procedure-statements FROM statement_index + 1 ASSIGNING FIELD-SYMBOL(<statement>).
+      DATA(tabix) = sy-tabix.
+      CASE <statement>-keyword.
+        WHEN 'ENDLOOP'.
+          DATA(end_idx) = sy-tabix.
```
Comment (bjoern-jueliger-sap) [inline src/checks/#cc4a#modern_language.clas.abap:578]: Declaring variable inside a scope and then later using it outside of it is confused, cf. also [style guide](https://github.com/SAP/styleguides/blob/main/clean-abap/CleanABAP.md#do-not-use-variables-outside-of-the-statement-block-they-are-declared-in)

File: src/checks/#cc4a#modern_language.clas.abap [hunk 8]
```diff
@@ -0,0 +1,1157 @@
...
+      CHANGING findings = findings ).
+      RETURN.
+    ENDIF.
+    DATA(exporting_idx) = analyzer->find_clause_index( tokens = statement-tokens clause = 'EXPORTING' ).
+
+*    IF exporting_idx <> 0.
+*      append_tokens( EXPORTING tokens = statement-tokens from_idx = exporting_idx + 1 CHANGING result = code_line ).
+*    ENDIF.
+
+*    code_line = |{ code_line } ).|.
```
Comment (bjoern-jueliger-sap) [inline src/checks/#cc4a#modern_language.clas.abap:708]: commented-out code

File: src/checks/#cc4a#modern_language.clas.abap [hunk 9]
```diff
@@ -0,0 +1,1157 @@
...
+                                   end_idx       = DATA(end_idx)
+                                   result_line   = DATA(result_line) ).
+
+    IF result_line IS INITIAL AND exporting_idxs IS INITIAL.
+      RETURN.
+    ENDIF.
+    DATA(quickfixes) = assistant_factory->create_quickfixes( ).
+    IF result_line IS NOT INITIAL.
+      DATA(finding_code) = c_code_exporting_receiving.
+      DATA(ps) = c_ps_exporting_receiving.
```
Comment (bjoern-jueliger-sap) [inline src/checks/#cc4a#modern_language.clas.abap:860]: Pspsps? I guess this is supposed to be `pseudo_comment` and not a method of attracting cats? ;)

File: src/checks/#cc4a#modern_language.clas.abap [hunk 10]
```diff
@@ -0,0 +1,1157 @@
...
+*              data(sep_by) = statement-tokens[ sy-tabix + 2 ]-lexeme.
+*              exit.
+*            when 'RESPECTING'.
+*              exit.
+*          endcase.
+*        endif.
+*        result_line = |{ result_line } { <token>-lexeme }|.
+*      endloop.
+*      result_line = |{ result_line } = |.
+*    endif.
```
Comment (bjoern-jueliger-sap) [inline src/checks/#cc4a#modern_language.clas.abap:922]: commented-out code

File: src/checks/#cc4a#modern_language.clas.abap [hunk 11]
```diff
@@ -0,0 +1,1157 @@
...
+       code = c_code_text_assembly
+       pseudo_comment = c_ps_text_assembly
+       quickfixes = quickfixes
+    CHANGING findings = findings ).
+
+  ENDMETHOD.
+
+
+  METHOD analyze_procedure.
+    "DATA(program_name) = cl_ci_atc_data_provider=>main_program_from_comp_unit( procedure-id-main_unit ).
```
Comment (bjoern-jueliger-sap) [inline src/checks/#cc4a#modern_language.clas.abap:992]: commented-out code

File: src/checks/#cc4a#modern_language.clas.abap [hunk 12]
```diff
@@ -0,0 +1,1157 @@
...
+        WHEN 'CREATE'.
+          IF <statement>-tokens[ 2 ]-lexeme = 'OBJECT' AND <statement>-tokens[ 2 ]-references IS INITIAL.
+            INSERT LINES OF analyze_create_object( procedure = procedure statement_index = idx ) INTO TABLE findings.
+          ENDIF.
+        WHEN 'CALL'.
+          IF <statement>-tokens[ 2 ]-lexeme = 'METHOD' AND <statement>-tokens[ 2 ]-references IS INITIAL.
+            INSERT LINES OF analyze_call_method( procedure = procedure statement_index = idx ) INTO TABLE findings.
+          ENDIF.
+*        WHEN 'CONCATENATE'.
+*          INSERT LINES OF analyze_text_assembly( procedure = procedure statement_index = idx ) INTO TABLE findings.
```
Comment (bjoern-jueliger-sap) [inline src/checks/#cc4a#modern_language.clas.abap:1014]: commented-out code

**Review by bjoern-jueliger-sap — CHANGES_REQUESTED** [review]

**Comment by D074020** [comment]
Issues seems to be solved, but do not merge this check until we have the current ABAP_ANALYZER version from Wieke. If you pull this check into a system without the current version, you will receive some syntax errors!

File: src/checks/#cc4a#modern_language.clas.xml [hunk 13]
```diff
@@ -0,0 +1,4 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<abapGit version="v1.0.0" serializer="LCL_OBJECT_CLAS" serializer_version="v1.0.0">
+ <asx:abap version="1.0" xmlns:asx="http://www.sap.com/abapxml"/>
+</abapGit>
```
Comment (bjoern-jueliger-sap) [inline src/checks/#cc4a#modern_language.clas.xml:4]: All metadata seems to be missing here (compare this to any other *.clas.xml file), and so I cannot import this into any other system to test it (error from Steampunk abapGit: "Invalid parameter OBJECT/OBJECTCLASS"). Maybe try to export this again?

**Review by bjoern-jueliger-sap — CHANGES_REQUESTED** [review]

File: src/checks/#cc4a#modern_language.clas.xml (hunk 13 above)
Comment (D074020) [reply src/checks/#cc4a#modern_language.clas.xml]: I have added the metadata, hope it does work now

**Review by D074020 — COMMENTED** [review]
Pseudo Review

**Review by bjoern-jueliger-sap — APPROVED** [review]
Can be merged for now because it works fine but might benefit from some refactoring in the longer methods later

**Pull request merged**